src/other/index.jl

```diff
@@ -61,6 +61,15 @@ rename!(f::Function, x::Index) = rename!(x, [(x=>f(x)) for x in x.names])
 rename(x::Index, args...) = rename!(copy(x), args...)
 rename(f::Function, x::Index) = rename!(f, copy(x))
 
+function permutecols!(x::Index, p::AbstractVector)
+    # to be callsed by permutecols! for DataFrame
```

Comment:
I'd rather call this `permute!` (just like I've kept `merge!`), as there's no ambiguity with rows vs. columns.

Also, instead of assuming that the permutation is valid, you can use `@boundscheck`, mark the function with `@propagate_inbounds`, and use `@inbounds` when calling it.

Finally, I think you could make this slightly faster by permuting `x.names`, and updating `x.lookup` in place.

Reply:
OK. In the process I thought then that we can make `Index` immutable by rewriting `names!` in a similar fashion.